File: src/gpx.ts
function escapeXml(value: string): string {
  return value
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&apos;");
}

/**
 * Render a GeoJSON LineString as a GPX 1.1 track.
 *
 * `coordinates` are `[lon, lat]` pairs or `[lon, lat, ele]` triples (GeoJSON
 * order); each becomes one `<trkpt lat lon>` with the order swapped. A nested
 * `<ele>` is emitted only for a tuple carrying a finite 3rd value. No `<time>`
 * is emitted.
 */
export function lineStringToGpx(
  coordinates: [number, number, number?][],
  name: string,
): string {
  const safeName = escapeXml(name);
  const points = coordinates
    .map(([lon, lat, ele]) =>
      typeof ele === "number" && Number.isFinite(ele)
        ? `      <trkpt lat="${lat}" lon="${lon}"><ele>${ele}</ele></trkpt>`
        : `      <trkpt lat="${lat}" lon="${lon}"/>`,
    )
    .join("\n");

  return `<?xml version="1.0" encoding="UTF-8"?>
<gpx version="1.1" creator="runna-router"
     xmlns="http://www.topografix.com/GPX/1/1"
     xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
     xsi:schemaLocation="http://www.topografix.com/GPX/1/1 http://www.topografix.com/GPX/1/1/gpx.xsd">
  <metadata><name>${safeName}</name></metadata>
  <trk>
    <name>${safeName}</name>
    <trkseg>
${points}
    </trkseg>
  </trk>
</gpx>
`;
}

File: ios/RouteKit/Sources/RouteKit/GPX.swift
import Foundation

/// GPX 1.1 track rendering, matching the server's `lineStringToGpx`.
public enum GPXWriter {
    public static func gpx(points: [RoutePoint], name: String) -> String {
        let safeName = escape(name)
        let body = points.map { point -> String in
            if let ele = point.elevation, ele.isFinite {
                return "      <trkpt lat=\"\(point.latitude)\" lon=\"\(point.longitude)\"><ele>\(ele)</ele></trkpt>"
            }
            return "      <trkpt lat=\"\(point.latitude)\" lon=\"\(point.longitude)\"/>"
        }.joined(separator: "\n")

        return """
            <?xml version="1.0" encoding="UTF-8"?>
            <gpx version="1.1" creator="loopr"
                 xmlns="http://www.topografix.com/GPX/1/1"
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                 xsi:schemaLocation="http://www.topografix.com/GPX/1/1 http://www.topografix.com/GPX/1/1/gpx.xsd">
              <metadata><name>\(safeName)</name></metadata>
              <trk>
                <name>\(safeName)</name>
                <trkseg>
            \(body)
                </trkseg>
              </trk>
            </gpx>

            """
    }

    private static func escape(_ value: String) -> String {
        value
            .replacingOccurrences(of: "&", with: "&amp;")
            .replacingOccurrences(of: "<", with: "&lt;")
            .replacingOccurrences(of: ">", with: "&gt;")
            .replacingOccurrences(of: "\"", with: "&quot;")
            .replacingOccurrences(of: "'", with: "&apos;")
    }
}

/// Reads `<trkpt>` points out of a GPX document.
public enum GPXParser {
    public static func points(in gpx: String) -> [RoutePoint] {
        guard let data = gpx.data(using: .utf8) else { return [] }
        let delegate = Delegate()
        let parser = XMLParser(data: data)
        parser.delegate = delegate
        parser.parse()
        return delegate.points
    }

    private final class Delegate: NSObject, XMLParserDelegate {
        var points: [RoutePoint] = []
        private var current: RoutePoint?
        private var inElevation = false
        private var text = ""

        func parser(
            _ parser: XMLParser, didStartElement name: String, namespaceURI: String?,
            qualifiedName: String?, attributes: [String: String] = [:]
        ) {
            switch name {
            case "trkpt":
                if let lat = attributes["lat"].flatMap(Double.init),
                    let lon = attributes["lon"].flatMap(Double.init)
                {
                    current = RoutePoint(longitude: lon, latitude: lat)
                }
            case "ele":
                inElevation = current != nil
                text = ""
            default:
                break
            }
        }

        func parser(_ parser: XMLParser, foundCharacters string: String) {
            if inElevation { text += string }
        }

        func parser(
            _ parser: XMLParser, didEndElement name: String, namespaceURI: String?,
            qualifiedName: String?
        ) {
            switch name {
            case "ele":
                current?.elevation = Double(text.trimmingCharacters(in: .whitespacesAndNewlines))
                inElevation = false
            case "trkpt":
                if let current { points.append(current) }
                current = nil
            default:
                break
            }
        }
    }
}
